ext/mbstring/libmbfl/mbfl/mbfl_encoding.h

```diff
@@ -129,10 +130,110 @@ struct mbfl_convert_vtbl {
 	void (*filter_copy)(struct _mbfl_convert_filter *src, struct _mbfl_convert_filter *dest);
 };
 
-/*
- * encoding
- */
-typedef struct _mbfl_encoding {
+typedef struct {
+	uint32_t state;
+	uint32_t errors;
+	zend_ulong maxsize;
+	size_t len;
+	char   val[1];
```

Comment:
Why not store a `zend_string*` in here? I don't really see the benefit of making mb_convert_bug itself resemble zend_string, and lots of disadvantages (no freedom to change the structure -- e.g. the maxsize here clearly should be size_t rather than zend_ulong, but can't be due to structure compatibility).

Reply:
Of course, in that case mb_convert_buf would be stack allocated, and we wouldn't return it either. Only the inner zend_string would be heap allocated.

Reply:
Let me give this a try. It's probably a better idea.

Reply:
Hmm. Implemented naively, this adds 20% perf overhead... youch.

Reply:
OK, well, it looks like the performance cost can be eliminated by caching the `zend_string` in a separate variable, so there's no extra pointer chasing every time we write to the output buffer.